ml-agents/mlagents/trainers/models.py

```diff
@@ -229,28 +230,101 @@ def create_visual_observation_encoder(
         :param h_size: Hidden layer size.
         :param activation: What type of activation function to use for layers.
         :param num_layers: number of hidden layers to create.
+        :param encoder_type: type of visual encoder (default/nature_cnn/resnet)
         :return: List of hidden layer tensors.
         """
         with tf.variable_scope(scope):
-            conv1 = tf.layers.conv2d(
-                image_input,
-                16,
-                kernel_size=[8, 8],
-                strides=[4, 4],
-                activation=tf.nn.elu,
-                reuse=reuse,
-                name="conv_1",
-            )
-            conv2 = tf.layers.conv2d(
-                conv1,
-                32,
-                kernel_size=[4, 4],
-                strides=[2, 2],
-                activation=tf.nn.elu,
-                reuse=reuse,
-                name="conv_2",
-            )
-            hidden = c_layers.flatten(conv2)
+            if encoder_type == "resnet":
+                n_channels = [16, 32, 32]  # channel for each stack
+                n_blocks = 2  # number of residual blocks
+                hidden = image_input
+                for i, ch in enumerate(n_channels):
+                    hidden = tf.layers.conv2d(
+                        hidden,
+                        ch,
+                        kernel_size=[3, 3],
+                        strides=[1, 1],
+                        reuse=reuse,
+                        name="layer%dconv_1" % i,
+                    )
+                    hidden = tf.layers.max_pooling2d(
+                        hidden, pool_size=[3, 3], strides=[2, 2], padding="same"
+                    )
+                    # create residual blocks
+                    for j in range(n_blocks):
+                        block_input = hidden
+                        hidden = tf.nn.relu(hidden)
+                        hidden = tf.layers.conv2d(
+                            hidden,
+                            ch,
+                            kernel_size=[3, 3],
+                            strides=[1, 1],
+                            padding="same",
+                            reuse=reuse,
+                            name="layer%d_%d_conv1" % (i, j),
+                        )
+                        hidden = tf.nn.relu(hidden)
+                        hidden = tf.layers.conv2d(
+                            hidden,
+                            ch,
+                            kernel_size=[3, 3],
+                            strides=[1, 1],
+                            padding="same",
+                            reuse=reuse,
+                            name="layer%d_%d_conv2" % (i, j),
+                        )
+                        hidden = tf.add(block_input, hidden)
+                hidden = tf.nn.relu(hidden)
+                hidden = c_layers.flatten(hidden)
+            elif encoder_type == "nature_cnn":
```

Comment:
I think in this case it would be better to put each of the different encoder types in its own `build_encoder` function -- so `build_nature_encoder` and `build_resnet_encoder` for example.

Reply:
yes sure. done.